nixos/modules/system/activation/activation-script.nix

```diff
@@ -200,6 +206,11 @@ in
   config = {
     system.activationScripts.stdio = ""; # obsolete
 
+    system.activationScripts.diff = lib.mkIf config.system.showSystemDiffOnRebuild ''
+      if [[ -e /run/current-system ]]; then
```

Comment:
I believe the [switch-to-configuration.pl](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/activation/switch-to-configuration.pl) script would be a better place for this. The activation script is run every time the system boots, but this diff only makes sense when you do an actual switch which is what `switch-to-configuration.pl` is for.

Reply:
`stc.pl` isn't really that hookable and I think it's better suited in the activation script since people tend to like it more (as they actually understand the language) and don't ask for it to be rewritten in another language all the time.
Probably also good for maintainability (as well as composability since we have this magic snippets thing and I could pretty trivially replace the thing if there's a differ I like more) to keep it here.

Reply:
I do not like pollution of the activation script phases. The point of the activation script is to activate a system derivation as atomically as possible. Too many phases are not good because they hurt atomicity (if a phase fails, there's nothing that rolls back previous phases unless you activate a previous generation), so they should be kept as small and atomic as possible. Dependencies between phases exist because certain things need to be set up before others, it's not really intended to be used by end users. In fact I think we already have too many builtin activation phases, a lot of them could be done using systemd units instead (with their obvious advantages).

While I can't see a critical issue with just a `diff` in an activation phase, I don't think the right place for it being written in Perl is a good justification for not putting it there. I'd be totally on board with rewriting `switch-to-configuration.pl` in e.g. Python btw.

Reply:
> The activation script is run every time the system boots, but this diff only makes sense when you do an actual switch which is what `switch-to-configuration.pl` is for.

I actually did a copy paste mistake. This should be userActivationScript.

I don't want add more code to switch-to-configuration because it is already 1000 lines long and having this in the module system is nicer because it also works when deploying different.

Reply:
Checking this again, `stc.pl` still makes more sense. In particular it would also allow `nixos-rebuild dry-activate` to show the diff. This couldn't be done with the activation script.

Yes the module system is nicer, but I'd also be onboard with making `stc.pl` be customizable via the module system.